spec/factories/order_factory.rb

```diff
@@ -113,6 +114,28 @@
                                        product: product)
       order.reload
     end
+
+    trait :completed do
+      transient do
+        completed_at { Time.zone.now }
+        state { "complete" }
+        payment_method { create(:payment_method, distributors: [distributor]) }
+        ship_address { create(:address) }
+      end
+
+      after(:create) do |order, evaluator|
+        # Ensure order is valid and passes through necessary checkout steps
+        create(:payment, state: "checkout", order: order, amount: order.total,
+                         payment_method: evaluator.payment_method)
+        order.ship_address = evaluator.ship_address
+        while !order.completed? do break unless order.next! end
```

Comment:
why don't we rather build the order and its associated objects in their final and valid state (like we do in the line below)? having to go through state machine (and hence, lots of callbacks) it's what makes our build slow to a crawl.

It's a trade-off, as always. We can't pretend to change things in models without updating our factories if that's what we want to avoid.

Reply:
There's various things that happen during the state machine steps that change the order and associated objects in various ways. There will be various changes being made to those state machine steps in the adjustments work, and it's nice to be able to have a factory option that puts the order through those steps (whatever they may be). There's quite a bit of complexity there...